File: packages/agents/src/structured-output.ts
import type { z } from "zod";

/**
 * Thrown when Agent.run()/resume() exhausts its repair attempts without the
 * model ever producing text that parses as JSON matching `responseSchema` —
 * `rawText` is the model's last (still-invalid) attempt, for a caller that
 * wants to log or salvage it rather than just seeing the thrown message.
 */
export class StructuredOutputError extends Error {
  constructor(
    message: string,
    readonly rawText: string,
  ) {
    super(message);
    this.name = "StructuredOutputError";
  }
}

export type StructuredOutputResult<T> = { success: true; data: T } | { success: false; error: string };

/**
 * Parses `text` as JSON and validates it against `schema` — the two failure
 * modes LangChain's `.with_structured_output()` repair loop feeds back to
 * the model (not valid JSON at all; valid JSON that doesn't match the
 * schema), collapsed into one human-readable `error` string suitable for
 * putting straight into a corrective prompt.
 */
export function parseStructuredOutput<T>(text: string, schema: z.ZodType<T>): StructuredOutputResult<T> {
  let json: unknown;
  try {
    json = JSON.parse(text);
  } catch (err) {
    return { success: false, error: `response is not valid JSON: ${err instanceof Error ? err.message : String(err)}` };
  }

  const result = schema.safeParse(json);
  if (result.success) {
    return { success: true, data: result.data };
  }

  return { success: false, error: formatZodIssues(result.error.issues) };
}

function formatZodIssues(issues: { path: PropertyKey[]; message: string }[]): string {
  return issues.map((issue) => `${issue.path.map(String).join(".") || "(root)"}: ${issue.message}`).join("; ");
}

/**
 * Zod's default ZodError.message is JSON.stringify(issues) — that's what a
 * tool's thrown ZodError carries whether it's an in-process Tool (a real
 * ZodError instance) or a resident-app export's input validation crossing
 * the RPC wire (already unwrapped to a plain Error(message) string by the
 * time it reaches Agent.run()'s catch block, per Computer's dispatch() — so
 * `instanceof ZodError` would never match the far more common resident-app
 * case). Detecting the shape of the *message string itself*, not the error
 * type, is what makes this work identically for either path, and for any
 * tool in any app, not one specific export — reformats it into the same
 * compact `path: message; path: message` shape parseStructuredOutput()
 * already produces, instead of leaving the model to parse a raw JSON array.
 * Any other error message (not a Zod issues array) passes through unchanged.
 */
export function formatToolInputError(message: string): string {
  let parsed: unknown;
  try {
    parsed = JSON.parse(message);
  } catch {
    return message;
  }

  if (!Array.isArray(parsed) || parsed.length === 0) return message;
  const looksLikeZodIssues = parsed.every(
    (issue) => issue && typeof issue === "object" && Array.isArray((issue as { path?: unknown }).path) && typeof (issue as { message?: unknown }).message === "string",
  );
  if (!looksLikeZodIssues) return message;

  return formatZodIssues(parsed as { path: PropertyKey[]; message: string }[]);
}

/** The corrective nudge fed back to the model as a fresh user turn on a failed attempt. */
export function structuredOutputRepairPrompt(error: string): string {
  return (
    `Your previous response could not be parsed as valid JSON matching the required schema:\n${error}\n\n` +
    `Respond again with ONLY corrected JSON matching the schema — no prose, no markdown code fences.`
  );
}

File: packages/agents-python/berth_agents/structured_output.py
"""Mirrors @berth/agents' structured-output.ts: parse a model's final answer
as JSON against a schema, and feed a corrective prompt back on failure.
Zod's schema/validation role is played by pydantic here — the closest
Python equivalent already used elsewhere in this repo (berth_sdk's own
manifest.py is pydantic-based too)."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ValidationError


class StructuredOutputError(Exception):
    """Raised when Agent.run()/resume() exhausts its repair attempts without
    the model ever producing text that parses as JSON matching
    `response_schema` — `raw_text` is the model's last (still-invalid)
    attempt, for a caller that wants to log or salvage it."""

    def __init__(self, message: str, raw_text: str) -> None:
        super().__init__(message)
        self.raw_text = raw_text


def parse_structured_output(text: str, schema: type[BaseModel]) -> tuple[bool, Any, str | None]:
    """Parses `text` as JSON and validates it against `schema` (a pydantic
    BaseModel subclass). Returns (success, data, error) — data is the
    validated model instance on success, error is a human-readable string
    on failure. pydantic's `model_validate_json` already collapses "not
    valid JSON at all" and "valid JSON that doesn't match the schema" into
    the same ValidationError type (a `json_invalid` issue for the former),
    so both failure modes flow through one except branch here."""
    try:
        data = schema.model_validate_json(text)
    except ValidationError as err:
        return False, None, _format_validation_error(err)
    return True, data, None


def _format_validation_error(err: ValidationError) -> str:
    parts = []
    for issue in err.errors():
        if issue["type"] == "json_invalid":
            parts.append(f"response is not valid JSON: {issue['msg']}")
            continue
        loc = ".".join(str(p) for p in issue["loc"]) or "(root)"
        parts.append(f"{loc}: {issue['msg']}")
    return "; ".join(parts)


def format_tool_input_error(err: BaseException) -> str:
    """Reformats a pydantic ValidationError raised by a Tool's own invoke()
    into the same compact per-field shape parse_structured_output() already
    produces, instead of leaving the model to parse pydantic's own multi-line
    error text. Unlike formatToolInputError() in the TypeScript package
    (which detects a Zod-issues-shaped JSON string, because a resident-app
    export's validation error crosses an RPC wire and arrives as a plain
    string), every Tool here runs in-process — there's no wire boundary to
    strip the exception type off, so checking `isinstance` directly is both
    simpler and reliable, not a shortcut. Any other exception's `str()` is
    returned unchanged."""
    if isinstance(err, ValidationError):
        return _format_validation_error(err)
    return str(err)


def structured_output_repair_prompt(error: str) -> str:
    """The corrective nudge fed back to the model as a fresh user turn on a
    failed attempt."""
    return (
        f"Your previous response could not be parsed as valid JSON matching the required schema:\n{error}\n\n"
        "Respond again with ONLY corrected JSON matching the schema — no prose, no markdown code fences."
    )
